sw/airborne/modules/calibration/mag_calib_ukf.c

```diff
@@ -0,0 +1,199 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/calibration/mag_calib_ukf.c"
+ * @author w.vlenterie
+ * Calibrate the magnetometer using an unscented kalman filter
+ * For more information please visit the following links:
+ *   - https://github.com/sfwa/trical
+ *   - http://au.tono.my/log/20131213-trical-magnetometer-calibration.html
+ *   - http://www.acsu.buffalo.edu/~johnc/mag_cal05.pdf
+ */
+
+#include <stdio.h>
+#include <error.h>
+#include <stdbool.h>
+#include <time.h>
+
+#include "modules/calibration/mag_calib_ukf.h"
+#include "subsystems/ahrs/ahrs_magnetic_field_model.h"
+#include "math/pprz_algebra_double.h"
+#include "TRICAL.h"
+#include "state.h"
+#include "modules/geo_mag/geo_mag.h"
+#include "generated/airframe.h"
+#include "subsystems/gps.h"
+#include "subsystems/abi_common.h"
+#include "abi_messages.h"
+
+static void mag_calib_ukf_run(uint8_t __attribute__((unused)) sender_id, uint32_t __attribute__((unused)) stamp, struct Int32Vect3 *mag);
+
+#define PRINT(string,...) fprintf(stderr, "[CALIB_UKF->%s()] " string,__FUNCTION__ , ##__VA_ARGS__)
+
+#if !defined MAG_CALIB_UKF_GEO_MAG_TIMEOUT
+#define MAG_CALIB_UKF_GEO_MAG_TIMEOUT 0
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_GEO_MAG_TIMEOUT)
+
+#if !defined MAG_CALIB_UKF_NORM
+#define MAG_CALIB_UKF_NORM 1.0f
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NORM)
+
+#if !defined MAG_CALIB_UKF_NOISE_RMS
+#define MAG_CALIB_UKF_NOISE_RMS 2e-1f
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_NOISE_RMS)
+
+#if !defined MAG_CALIB_UKF_HOTSTART
+#define MAG_CALIB_UKF_HOTSTART TRUE
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART)
+
+#if !defined MAG_CALIB_UKF_HOTSTART_SAVE_FILE
+#define MAG_CALIB_UKF_HOTSTART_SAVE_FILE /data/ftp/internal_000/mag_ukf_calib.txt
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_HOTSTART_SAVE_FILE)
+
+#if !defined MAG_CALIB_UKF_VERBOSE || !MAG_CALIB_UKF_VERBOSE
+#define VERBOSE_PRINT(...)
+#else
+#define VERBOSE_PRINT PRINT
+#endif
+PRINT_CONFIG_VAR(MAG_CALIB_UKF_VERBOSE)
+
+TRICAL_instance_t mag_calib;
+static abi_event mag_ev;
+
+time_t mag_field_from_geo_mag;
+bool   update_geo_mag_field = true;
+struct FloatVect3 H = { .x = MAG_CALIB_UKF_NORM, .y = 0.0f, .z =  0.0f};
+
+#if MAG_CALIB_UKF_HOTSTART
+static FILE *fp;
+char hotstart_file_name[512];
+#endif
+
+void mag_calib_ukf_init(void)
+{
+#if USE_MAGNETOMETER
+  TRICAL_init(&mag_calib);
+  TRICAL_norm_set(&mag_calib, MAG_CALIB_UKF_NORM);
+  TRICAL_noise_set(&mag_calib, MAG_CALIB_UKF_NOISE_RMS);
+
+  H.x = AHRS_H_X;
+  H.y = AHRS_H_Y;
+  H.z = AHRS_H_Z;
+  VERBOSE_PRINT("Local magnetic field loaded from airframe file (Hx: %4.2f, Hy: %4.2f, Hz: %4.2f)\n", H.x, H.y, H.z);
+
+  mag_calib_hotstart_read();
+  AbiBindMsgIMU_MAG_INT32(IMU_BOARD_ID, &mag_ev, mag_calib_ukf_run);
```

Comment:
The ID should be something like `MAG_CALIB_UKF_MAG_ID` with a default to:
```
#ifndef MAG_CALIB_UKF_MAG_ID
#define MAG_CALIB_UKF_MAG_ID ABI_BROADCAST
#endif
```
This way it can be used with and external mag or binded to a specific one when multiple sensors are present.